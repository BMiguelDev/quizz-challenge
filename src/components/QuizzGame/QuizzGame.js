import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';

import Question from '../Question/Question';
import styles from './QuizzGame.module.scss';


export default function QuizzGame({ isLoading, quizzQuestions, handleSelectAnswer, quizzResults, handleCheckResults, handleQuizzReset }) {

    // Constant holding the completion bar filling styles, which varry based on the number of corret answers
    const borderFillingStyles = {
        width: quizzResults[1] > 0 ? quizzResults[1] * 20 + "%" : "10%",
        padding: quizzResults[1] > 0 ? ".7rem" : "0.7rem .2rem",
        border: quizzResults[1] === 0 && "none",
        left: quizzResults[1] === 0 && "0%",
    }

    return (
        <div className={styles.quizz_container}>
            {
                isLoading ?
                    <div className={styles.loading_container}>
                        <FontAwesomeIcon icon={faSpinner} />
                    </div> :
                    <div className={styles.questions_container} aria-label='questions_container'>
                        {quizzQuestions.map((questionData, index) =>
                            <Question
                                key={index}
                                questionData={questionData}
                                questionIndex={index}
                                handleSelectAnswer={handleSelectAnswer}
                                isShowResults={quizzResults[0]} />)}
                    </div>
            }
            <div className={styles.results_container}>
                <div className={quizzResults[0] === false ? `${styles.results_description} ${styles.results_description_hide}` : styles.results_description} aria-label="results_description">
                    <h3 className={styles.results_correct_answers}>You scored {quizzResults[1]}/5 correct answers</h3>
                    <span className={styles.completion_bar_frame}><span style={borderFillingStyles} className={styles.completion_bar_filling}></span></span>
                </div>
                <div className={styles.results_buttons_container}>
                    <button className={styles.quizz_action_button} onClick={handleCheckResults}>{quizzResults[0] === true ? "Play Again" : "Check Answers"}</button>
                    <button className={styles.quizz_reset_button} onClick={handleQuizzReset}>Reset Quizz</button>
                </div>
            </div>
        </div>
    );
}
