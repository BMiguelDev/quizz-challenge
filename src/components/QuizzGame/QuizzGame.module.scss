// Quizz content
.quizz_container {
    // padding: 4rem 6rem;
    padding: 3% 7% 1.5% 7%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-evenly;
    width: 100%;
    height: 92%;
    gap: 1.5%;

    .loading_container {
        width: 100%;
        display: flex;
        flex: 2.5;
        align-items: center;
        justify-content: center;


        svg {
            color: var(--main-text-color);
            font-size: 6rem;

            @keyframes spinInfinite {
                0% {
                    //rotate: x 90deg;

                    rotate: 0;
                }

                100% { 
                    rotate: 360deg;
                }
            }

            animation: spinInfinite 1.25s linear 0s infinite;
        
        }
    }
}


// Question content
.questions_container {
    display: flex;
    flex-direction: column;
    gap: 1.15rem;
    overflow-y: auto;
    min-width: 70%;
    max-width: 95%;
    padding: .25rem 2rem;
}


// Results content
.results_container {
    display: flex;
    width: 100%;
    align-items: center;
    justify-content: center;
    gap: 2.5rem;
    padding-top: 1rem;

    .results_description {
        display: flex;
        flex-direction: column;
        gap: .5rem;

        &.results_description_hide {
            display: none;
        }

        .results_correct_answers {
            margin: 0;
            color: var(--main-text-color);
        }

        .completion_bar_frame {
            padding: .7rem;
            border: .08rem solid var(--completion-bar-outer-border-bg-color);
            border-radius: 1rem;
            background-color: var(--completion-bar-bg-color);
            position: relative;

            .completion_bar_filling {
                position: absolute;
                top: 0;
                left: -.5%;
                padding: .7rem;
                background-color: transparent;
                transition: all .3s ease-in-out;

                &::before {
                    content: '';
                    position: absolute;
                    width: 100%;
                    height: 100%;
                    background-color: var(--completion-bar-filling-bg-color);
                    top: -3.5%;
                    left: -.5%;
                    border-radius: 1rem;
                    border: .01rem solid var(--completion-bar-filling-border-color);

                    @keyframes completionBarAnimation {
                        0% {
                            width: 0%;
                            opacity: 0;
                        }

                        50% {
                            width: 35%;
                            opacity: .5;
                        }

                        100% {
                            width: 100%;
                            opacity: 1;
                        }
                    }

                    animation: completionBarAnimation 1.2s linear;
                }
            }
        }
    }

    .results_buttons_container {
        display: flex;
        align-items: center;
        justify-content: center;
        gap: 2rem;
        
        .quizz_action_button {
            background-color: var(--button-bg-color);
            color: var(--button-color);
            padding: .75rem 1rem;
            border-radius: .75rem;
            font-size: 1rem;
            border: none;
            cursor: pointer;
            font-family: 'Roboto', sans-serif;
            transition: all .2s ease-in-out;
        }

        .quizz_reset_button {
            background-color: var(--quizz-reset-button-bg-color);
            color: var(--button-color);
            padding: .75rem 1rem;
            border-radius: .75rem;
            font-size: 1rem;
            border: none;
            cursor: pointer;
            font-family: 'Roboto', sans-serif;
            transition: all .2s ease-in-out;
        }
    }
}


@import './media';
