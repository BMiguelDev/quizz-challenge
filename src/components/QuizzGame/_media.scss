// Media query for mobile extremelly small screens
@media screen and (min-width: 250px) and (max-width: 400px) and (min-height: 200px),
screen and (min-height: 200px) and (max-height: 450px) and (min-width: 250px) {

    // Quizz content
    .quizz_container {
        padding: .25rem 1rem;
        padding-top: 1rem;
        gap: .5rem;
        justify-content: center;
        height: 94%;

        .loading_container {
            flex: .65;

            svg {
                font-size: 3rem;
            }
        }
    }


    // Questions content
    .questions_container {
        gap: .5rem;
        width: 100%;
        height: 82.5%;
        // overflow-y: scroll;
        overflow-y: auto;
        max-width: unset;
        min-width: unset;
        padding: .2rem 1.5rem;
    }

    
    // Results content
    .results_container {
        gap: 1.25rem;
        max-height: 17.5%;
        padding: .15rem 1.5rem .2rem 1.5rem;

        .results_description {
            flex: 2;
            gap: .25rem;

            .results_correct_answers {
                font-size: .575rem;
                line-height: .55rem;
                width: 90%;
            }

            .completion_bar_frame {
                padding: .175rem;
                width: 90%;

                .completion_bar_filling {

                    &::before {
                        height: 25%;
                    }
                }
            }
        }

        .results_buttons_container {
            gap: .75rem;
            flex: 2.75;

            .quizz_action_button {
                padding: .3rem .5rem;
                font-size: .55rem;
                line-height: .6rem;
            }

            .quizz_reset_button {
                padding: .3rem .5rem;
                font-size: .55rem;
                line-height: .6rem;
            }
        }
    }
}


// Media query for mobile content (portrait)
@media screen and (min-width: 400px) and (max-width: 550px) and (min-height: 450px) {

    // Quizz content
    .quizz_container {
        padding: .5rem 1.5rem;
        padding-top: 3.5rem;
        gap: 1rem;
        justify-content: center;

        .loading_container {
            flex: .65;

            svg {
                font-size: 4.5rem;
            }
        }
    }


    // Questions content
    .questions_container {
        gap: 1rem;
        width: 100%;
        height: 90%;
        // overflow-y: scroll;
        overflow-y: auto;
        max-width: unset;
        min-width: unset;
        padding: .225rem 1.5rem;
    }


    // Results content
    .results_container {
        gap: 1.25rem;
        height: 10%;
        padding: .25rem 0 .25rem 0;
        padding: .25rem 1.5rem;

        .results_description {
            flex: 2;
            gap: .5rem;

            .results_correct_answers {
                font-size: .775rem;
                line-height: .6rem;
            }

            .completion_bar_frame {
                padding: .35rem;

                .completion_bar_filling {

                    &::before {
                        height: 50%;
                    }
                }
            }
        }

        .results_buttons_container {
            gap: 1.25rem;
            flex: 2.5;

            .quizz_action_button {
                padding: .35rem .7rem;
                font-size: .7rem;
            }

            .quizz_reset_button {
                padding: .35rem .7rem;
                font-size: .7rem;
            }
        }
    }
}


// Media query for mobile content (landscape)
@media screen and (min-height: 450px) and (max-height: 900px) and (min-width: 550px) and (max-width: 900px),
screen and (min-height: 900px) and (min-width: 550px) and (max-width: 650px),
screen and (min-height: 450px) and (max-height: 650px) and (min-width: 900px) {

     // Quizz content
     .quizz_container {
        padding: 1.25rem 3.5rem;
        padding-bottom: .75rem;

        .loading_container {
            flex: .75;

            svg {
                font-size: 5rem;
            }
        }
    }


    // Questions content
    .questions_container {
        overflow-x: hidden;
        // overflow-y: scroll;
        overflow-y: auto;
        width: 100%;
        height: 80%;
        max-width: unset;
        min-width: unset;
    }


    // Results content
    .results_container {
        width: 100%;
        gap: 1.5rem;
        padding-top: 1rem;
        position: relative;
        height: 18%;
        border-top: .01rem solid var(--button-bg-color);
        padding: .5rem 0 .15rem 0;
        align-items: center;

        .results_description {
            flex: 2.75;
            gap: .35rem;

            .results_correct_answers {
                font-size: .85rem;
                text-align: center;
            }

            .completion_bar_frame {
                padding: .4rem;

                .completion_bar_filling {

                    &::before {
                        height: 52.5%;
                    }
                }
            }
        }

        .results_buttons_container {
            gap: 1.25rem;
            flex: 2.5;

            .quizz_action_button {
                padding: .25rem .8rem;
                font-size: .725rem;
            }

            .quizz_reset_button {
                padding: .25rem .8rem;
                font-size: .725rem;
            }
        }
    }
}
