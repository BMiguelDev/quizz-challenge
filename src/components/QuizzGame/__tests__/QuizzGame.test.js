import { cleanup, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";

import QuizzGame from "../QuizzGame";


jest.mock("nanoid", () => ({
    __esModule: true,
    nanoid: () => Date.now()
}))


afterEach(() => {
    cleanup();
})


const mockFunction = jest.fn();
const quizzQuestionsTest = [
    {
        category: "Entertainment: Video Games",
        correct_answer: "The Zin Empire",
        difficulty: "easy",
        incorrect_answers: ['The Brotherhood', 'The Ronin', 'The Sons of Samedi'],
        question: "Which of these is NOT the name of a rival gang in the video game Saint&#039;s Row 2?",
        type: "multiple"
    },
    {
        category: "Science: Computers",
        correct_answer: "False",
        difficulty: "medium",
        incorrect_answers: ['True'],
        question: "The common software-programming acronym &quot;I18N&quot; comes from the term &quot;Interlocalization&quot;.",
        type: "boolean"
    },
    {
        category: "Entertainment: Music",
        correct_answer: "Pink Floyd",
        difficulty: "medium",
        incorrect_answers: ['AC/DC', 'Metallica', 'Red Hot Chili Peppers'],
        question: "Which of these bands is the oldest?",
        type: "multiple"
    },
    {
        category: "Entertainment: Television",
        correct_answer: "LazyTown",
        difficulty: "easy",
        incorrect_answers: ['Sofia the First', 'DuckTales', 'Tom and Jerry'],
        question: "Which show is known for the songs &quot;You are a Pirate&quot;, &quot;Cooking by the Book&quot; and &quot;We Are Number One&quot;?",
        type: "multiple"
    },
    {
        category: "Entertainment: Video Games",
        correct_answer: "Shi No Numa",
        difficulty: "medium",
        incorrect_answers: ['Tranzit', 'Kino Der Toten', 'Der Riese'],
        question: "Which was the first &quot;Call Of Duty: Zombies&quot; map to introduce the &quot;Wunderwaffe DG-2&quot;?",
        type: "multiple"
    }
]


test('component should always match snaphot', () => {
    const tree = renderer.create(<QuizzGame isLoading={false} quizzQuestions={quizzQuestionsTest} handleSelectAnswer={mockFunction} quizzResults={[false, 0]} handleCheckResults={mockFunction} handleQuizzReset={mockFunction} />);

    expect(tree).toMatchSnapshot();
})


test('component should render 5 questions', () => {
    render(<QuizzGame isLoading={false} quizzQuestions={quizzQuestionsTest} handleSelectAnswer={mockFunction} quizzResults={[false, 0]} handleCheckResults={mockFunction} handleQuizzReset={mockFunction} />);

    const questionsElement = screen.getByTitle("questions_container");

    expect(questionsElement.childElementCount).toEqual(5);
})


test("score should not be visible if quizz is not finished", () => {
    render(<QuizzGame isLoading={false} quizzQuestions={quizzQuestionsTest} handleSelectAnswer={mockFunction} quizzResults={[false, 0]} handleCheckResults={mockFunction} handleQuizzReset={mockFunction} />);

    const resultsDescriptionElement = screen.getByTitle("results_description");

    expect(resultsDescriptionElement.classList).toContain("results_description_hide");
})


test('both "Check Answers" and "Reset Quizz" buttons should be enabled if quizz is not finished', () => {
    render(<QuizzGame isLoading={false} quizzQuestions={quizzQuestionsTest} handleSelectAnswer={mockFunction} quizzResults={[false, 0]} handleCheckResults={mockFunction} handleQuizzReset={mockFunction} />);

    const buttonElement1 = screen.getByText("Check Answers");
    const buttonElement2 = screen.getByText("Reset Quizz");

    expect(buttonElement1).toBeEnabled();
    expect(buttonElement2).toBeEnabled();
})

test('both "Play Again" and "Reset Quizz" buttons should be enabled if quizz is finished', () => {
    render(<QuizzGame isLoading={false} quizzQuestions={quizzQuestionsTest} handleSelectAnswer={mockFunction} quizzResults={[true, 0]} handleCheckResults={mockFunction} handleQuizzReset={mockFunction} />);

    const buttonElement1 = screen.getByText("Play Again");
    const buttonElement2 = screen.getByText("Reset Quizz");

    expect(buttonElement1).toBeEnabled();
    expect(buttonElement2).toBeEnabled();
})

test('component should show number of correct answers if quizz is finished', () => {
    const numberCorrectAnswers = 3;
    render(<QuizzGame isLoading={false} quizzQuestions={quizzQuestionsTest} handleSelectAnswer={mockFunction} quizzResults={[true, numberCorrectAnswers]} handleCheckResults={mockFunction} handleQuizzReset={mockFunction} />);

    const headingElements = screen.getAllByRole("heading");
    const resultsScoreHeadingElement = headingElements[headingElements.length-1];

    expect(resultsScoreHeadingElement).toHaveTextContent(`You scored ${numberCorrectAnswers}/5 correct answers`);
})
