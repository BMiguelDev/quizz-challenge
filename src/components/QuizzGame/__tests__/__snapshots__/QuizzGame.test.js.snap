// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component should always match snaphot 1`] = `
<div
  className="quizz_container"
>
  <div
    aria-label="questions_container"
    className="questions_container"
  >
    <div
      className="question_container"
    >
      <h3
        className="question"
      >
        Which of these is NOT the name of a rival gang in the video game Saint's Row 2?
      </h3>
      <div
        className="answers_container"
      />
      <hr
        className="question_separator"
      />
    </div>
    <div
      className="question_container"
    >
      <h3
        className="question"
      >
        The common software-programming acronym "I18N" comes from the term "Interlocalization".
      </h3>
      <div
        className="answers_container"
      />
      <hr
        className="question_separator"
      />
    </div>
    <div
      className="question_container"
    >
      <h3
        className="question"
      >
        Which of these bands is the oldest?
      </h3>
      <div
        className="answers_container"
      />
      <hr
        className="question_separator"
      />
    </div>
    <div
      className="question_container"
    >
      <h3
        className="question"
      >
        Which show is known for the songs "You are a Pirate", "Cooking by the Book" and "We Are Number One"?
      </h3>
      <div
        className="answers_container"
      />
      <hr
        className="question_separator"
      />
    </div>
    <div
      className="question_container"
    >
      <h3
        className="question"
      >
        Which was the first "Call Of Duty: Zombies" map to introduce the "Wunderwaffe DG-2"?
      </h3>
      <div
        className="answers_container"
      />
      <hr
        className="question_separator"
      />
    </div>
  </div>
  <div
    className="results_container"
  >
    <div
      aria-label="results_description"
      className="results_description results_description_hide"
    >
      <h3
        className="results_correct_answers"
      >
        You scored 
        0
        /5 correct answers
      </h3>
      <span
        className="completion_bar_frame"
      >
        <span
          className="completion_bar_filling"
          style={
            Object {
              "border": "none",
              "left": "0%",
              "padding": "0.7rem .2rem",
              "width": "10%",
            }
          }
        />
      </span>
    </div>
    <div
      className="results_buttons_container"
    >
      <button
        className="quizz_action_button"
        onClick={[MockFunction]}
      >
        Check Answers
      </button>
      <button
        className="quizz_reset_button"
        onClick={[MockFunction]}
      >
        Reset Quizz
      </button>
    </div>
  </div>
</div>
`;
